app/controllers/shipit/api/locks_controller.rb

```diff
@@ -19,14 +19,29 @@ def create
         requires :reason, String, presence: true
       end
       def update
-        stack.update(lock_reason: params.reason, lock_author: current_user)
+        if stack.locked?
+          stack.update(lock_reason: params.reason, lock_author: current_user)
+        else
+          lock(params.reason, current_user)
+        end
         render_resource stack
       end
 
       def destroy
-        stack.update(lock_reason: nil, lock_author: nil)
+        unlock
         render_resource stack
       end
+
+      private
+
+      def lock(reason, user)
+        stack.update(lock_reason: reason, lock_author: user,
+                     locked_since: Time.current)
```

Comment:
Oh, I see you handle it in the controller. It would be better if lock would handle this transparently.